js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/useShippingRatesWithSavedSuggestions.js

```diff
@@ -0,0 +1,97 @@
+/**
+ * External dependencies
+ */
+import { useState, useRef, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useShippingRates from '.~/hooks/useShippingRates';
+import useShippingRatesSuggestions from './useShippingRatesSuggestions';
+import useSaveSuggestions from './useSaveSuggestions';
+import useCallbackOnceEffect from '.~/hooks/useCallbackOnceEffect';
+
+/**
+ * @typedef {Object} ShippingRatesWithSavedSuggestionsResult
+ * @property {boolean} loading Whether loading is in progress.
+ * @property {Array<import('.~/data/actions').ShippingRate>?} data Shipping rates.
+ */
+
+/**
+ * Check existing shipping rates, and if it is empty, get shipping rates suggestions
+ * and save the suggestions as shipping rates.
+ *
+ * @return {ShippingRatesWithSavedSuggestionsResult} Result object with `loading` and `data`.
+ */
+const useShippingRatesWithSavedSuggestions = () => {
+	const {
+		hasFinishedResolution: hfrShippingRates,
+		data: dataShippingRates,
+	} = useShippingRates();
+	const {
+		loading: loadingSuggestions,
+		data: dataSuggestions,
+	} = useShippingRatesSuggestions();
+
+	/**
+	 * `isInitialShippingRatesEmpty` is used to indicate
+	 * whether the initial loaded shipping rates
+	 * has a pre-saved value or not.
+	 *
+	 * If it does not have a pre-saved value,
+	 * shipping rates should be an empty array,
+	 * and we should save the suggestions as shipping rates.
+	 *
+	 * If it does have a pre-saved value,
+	 * then we should not save the suggestions,
+	 * even when users manually deleted all the pre-saved shipping rates value.
+	 * The exception is when users deleted all the pre-saved value
+	 * and then reload the page,
+	 * then the suggestions would be saved as shipping rates as per above logic.
+	 */
+	const isInitialShippingRatesEmpty = useRef( undefined );
+	if (
+		hfrShippingRates &&
+		isInitialShippingRatesEmpty.current === undefined
+	) {
+		isInitialShippingRatesEmpty.current = dataShippingRates.length === 0;
+	}
```

Comment:
💅 

As mentioned, for clarity I would save this logic in a wrap like 

`const shouldSaveSuggestions = ... `

Reply:
@puntope , could you elaborate more? How should I "save this logic in a wrap like `const shouldSaveSuggestions = ...`"? 🤔

Reply:
I've changed the namings for better readability in [`b0f6a9a` (#1161)](https://github.com/woocommerce/google-listings-and-ads/pull/1161/commits/b0f6a9afed22ba5e780d4beed3aea04932475d14).